### PR TITLE
Parameterizable and type safe routing rework

### LINKED ISSUE #17: move from generic to protocol based type safe routing
Type safe routing is currently limited to 3 methods. 

```swift
// doesn't work
drop.get("users", User.self, "posts", Post.self) { req, user, post in
    return "User #\(user.id) and post #(post.id)"
}
```

This is due to limitations with variadics and generics. 

I propose moving to a protocol-based approach.

```swift
drop.get("users", User.parameter, "posts", Post.parameter) { req in
    let user = try req.parameters.get(User.self)
    let user = try req.parameters.get(User.self)
    
    return "User #\(user.id) and post #(post.id)"
}
```

This could be implemented as a `Parameterizable` protocol. Model would conform to this by default. 

```swift
public protocol Parameterizable {
    // the unique key to use as a slug in route building
    static var uniqueSlug: String { get }

    // returns the found model for the resolved url parameter
    static func make(for parameter: String) throws -> Self
}

extension Parameterizable {
    static var parameter: String {
        return ":" + uniqueParameterKey
    }
}
```

```swift
extension Parameters {
    func get<P: Parameterizable>(_ p: P.Type = P.self) throws -> P {
         let parameter = try get(P.uniqueSlug) as String
         return try P.make(for: parameter)
    }
}
```

### PR DESCRIPTION
**Pull request by BrettRToomey**
Fixes #17 and #16

### CODE REVIEW

File: Sources/Routing/Parameterizable.swift [hunk 1]
```diff
@@ -0,0 +1,13 @@
+public protocol Parameterizable {
+    /// the unique key to use as a slug in route building
+    static var uniqueSlug: String { get }
+    
+    // returns the found model for the resolved url parameter
+    static func make(for parameter: String) throws -> Self
+}
+
+extension Parameterizable {
+    public static var parameter: String {
```
Comment (tanner0101) [inline Sources/Routing/Parameterizable.swift]: Add a comment here about how this should be used in routing methods.

File: Sources/Routing/RouteBuilder.swift [hunk 2]
```diff
@@ -28,10 +30,56 @@ extension RouteBuilder {
...
+        }
+    }
+    
+    public func delete(_ segments: String..., handler: @escaping RouteHandler) {
+        register(method: .delete, path: segments) {
+            try handler($0).makeResponse()
+        }
+    }
+    
+    public func options(_ segments: String..., handler: @escaping RouteHandler) {
```
Comment (tanner0101) [inline Sources/Routing/RouteBuilder.swift]: We should have the generic `add` function here as well as an `all` function (should work by just adding `*` but i can't remember).

**Comment by Evertt** [comment]
I prefer the current way. First of all it forces you to organize your routes well using route groups or classes and more importantly I find `.self` more beautiful than `.parameter`. And I hope that in future swift versions we can even get rid of `.self`.

**Comment by BrettRToomey** [comment]
The old way is generic and has hardcoded overloads for parameters ranging from 0, to 3. That implementation severe ramifications including:
- You can only have 3 or less path parameters
- Bloated binary
- Heavy build times
- Heavy start up (negligible in most cases)
- Loss of useful error messages. When generics are involved, especially hacky-ish generics like these, error message can be extremely misleading if not useless.

I really appreciate the convenience of having my parameters passed back to me in my closure, but too much is sacrificed. Maybe in the future Swift will have some stronger reflection or some ways to better accomplish our goals, but at this moment I feel like this is a choice we have to make.

File: Sources/Routing/Parameters.swift [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    public let context: Context
+    
+    public init(_ wrapped: StructuredData, in context: Context? = defaultContext) {
+        self.wrapped = wrapped
+        self.context = context ?? parametersContext
+    }
+}
+
+extension Parameters {
+    public func get<P: Parameterizable>(_ p: P.Type = P.self) throws -> P {
```
Comment (LoganWright) [inline Sources/Routing/Parameters.swift]: Have we verified that we don't get ambiguous warnings outside of this module when using this if an object conforms to Parameterizable _and_ NodeInitializable

File: Sources/Routing/Parameterizable.swift [hunk 4]
```diff
@@ -0,0 +1,25 @@
+public protocol Parameterizable {
+    /// the unique key to use as a slug in route building
+    static var uniqueSlug: String { get }
```
Comment (LoganWright) [inline Sources/Routing/Parameterizable.swift]: Having a static variable here, is it possible to have the same type more than once, ie: `drop.get(String, "foo", String)`

Comment (LoganWright) [inline Sources/Routing/Parameterizable.swift]: this has been resolved, and duplicate param names are now acceptable 👍

**Pull request merged**